File: arrays/diagonalTraverse.py
# https://leetcode.com/problems/diagonal-traverse/

from typing import List
from collections import defaultdict

class Solution:
    
    # Runtime: 347 ms, faster than 28.51% of Python3 online submissions for Diagonal Traverse.
    # Memory Usage: 17.2 MB, less than 78.68% of Python3 online submissions for Diagonal Traverse.
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        
        dictionary = defaultdict(list)
        R, C  = len(mat), len(mat[0])
        
        
        for i in range (R):
            for j in range(C):
                dictionary[i+j].append(mat[i][j])

        result = []    
        for i in range (R-1 +  C-1 + 1):
            if i&1==0:
                dictionary[i].reverse()
            
            result.extend(dictionary[i])
        return result
    
    
    # Runtime: 214 ms, faster than 83.03% of Python3 online submissions for Diagonal Traverse.
    # Memory Usage: 17.1 MB, less than 90.31% of Python3 online submissions for Diagonal Traverse.
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        endrow, endcol = len(mat), len(mat[0])
        
        def go_up(row, col):
            result = []
            while 0<=row<endrow and 0<=col<endcol:
                result.append(mat[row][col])
                row-=1
                col+=1
            return result
        

        result = []
        
        for i in range(endrow):
            result.append(go_up(i, 0))
            
        for i in range (1, endcol):
            result.append(go_up(endrow-1, i))
            
        answer = []
        for ind , ls in enumerate (result):
            answer.extend(ls[::-1] if ind&1 else ls)
        return answer
            

File: arrays/diagonalTraverse.js
/**
 * @param {number[][]} mat
 * @return {number[]}
 */
const findDiagonalOrder = (mat) => {
    let rows = mat.length;
    let cols = mat[0].length;
    let result = new Array(rows + cols - 1).fill(null).map(() => []);

    for (let row = 0; row < rows; row++) {
        for (let col = 0; col < cols; col++) {
            if ((row + col) % 2 === 0) result[row + col].unshift(mat[row][col]);
            else result[row + col].push(mat[row][col]);
        }
    }
    return result.flat();
};
module.exports = findDiagonalOrder([[1, 2, 3], [4, 5, 6], [7, 8, 9]]);